File: src/c/kamada_kawai/algo/algo.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>
#include <omp.h>
#include <time.h>
#include <signal.h>
#include <unistd.h>
#include <errno.h>

#include "../kamada_kawai.h"
#include "algo.h"

void output_algo(KamadaKawai *kk, Coord *original, double avg_time, char *name, char *test_id)
{
    // --- WRITE RESULTS TO FILE --- //
    char *algoresult = generateFilename(name, test_id, "csv");
    FILE *algoresult_fp = fopen(algoresult, "w");
    if (algoresult_fp == NULL)
    {
        fprintf(stderr, "Error opening file %s: %s\n", algoresult, strerror(errno));
        return;
    }
    writeState(algoresult_fp, original, kk->n);
    writeState(algoresult_fp, kk->coords, kk->n);
    fclose(algoresult_fp);
    // --- WRITE TIME TO FILE --- //
    char *algotime = generateFilename(name, test_id, "txt");
    FILE *algotime_fp = fopen(algotime, "w");
    if (algotime_fp == NULL)
    {
        fprintf(stderr, "Error opening file %s: %s\n", algotime, strerror(errno));
        return;
    }
    writeTime(algotime_fp, avg_time);
    fclose(algotime_fp);
}

void run_algo(KamadaKawai *kk, void (*func)(KamadaKawai *), char *name, char *test_id, char *repetitions)
{
    // --- SETUP --- //
    Coord *original = malloc(kk->n * sizeof(Coord));
    copyCoords(kk->coords, original, kk->n); // Copy the original data

    // --- RUN ALGORITHM R times --- //
    double avg_time = 0;
    for (int i = 0; i < atoi(repetitions); i++)
    {
        // --- RESET DATA --- //
        copyCoords(original, kk->coords, kk->n); // Copy the original data back

        // --- EXECUTION --- //
        double start = omp_get_wtime(); // Start algorithm time
        func(kk);                       // Run the algorithm
        double end = omp_get_wtime();   // End algoritYYYm time
        avg_time += end - start;
    }
    avg_time = avg_time / atoi(repetitions);
    
    // --- OUTPUT --- //
    output_algo(kk, original, avg_time, name, test_id); // Write results to file
    
    // --- RESET DATA --- //
    copyCoords(original, kk->coords, kk->n); // Copy the original data back

    free(original); // Free the original data
    return;
}

void algo(KamadaKawai *kk, char *test_id, char *repetitions)
{
    // run_algo(kk, mem_seq, "mem_seq", test_id, repetitions);
    // run_algo(kk, mem_par, "mem_par", test_id, repetitions);
    run_algo(kk, par, "par", test_id, repetitions);

    return;
}

File: src/julia/partition/algo/algo.jl
module Algo

include("./seq.jl")
using .Seq
include("./sgl_stc.jl")
using .Sgl_stc
include("../partitions2json.jl")
using .Partitions2Json
include("../../common/write.jl")
using .Write
include("../../common/file.jl")
using .File

using BenchmarkTools

export algo

function output_algo(result::Vector{Bool}, elapsed, name::String, test_id::String)
	# --- WRITE RESULTS TO FILE --- #
	algoresult = generateFilename(name, test_id, "json")
	writePartitions(algoresult, result)
	# --- WRITE TIME TO FILE --- #
	algotime = generateFilename(name, test_id, "txt")
	writeTime(algotime, elapsed)
end

function run_algo(arr::Vector{Vector{Int64}}, f::Function, name::String, test_id::String, repetitions::Int)
    # --- SETUP --- #
    result = Vector{Bool}(undef, length(arr))

    # --- BENCHMARK --- #
	GC.gc() # Force garbage collection to ensure memory is cleaned up before output
	bench = @benchmark $f($arr) samples=repetitions evals=1

    # Get median execution time in seconds
    avg_time = median(bench).time / 1e9

    # Run once to get the actual result to save
    result = f(arr)

    output_algo(result, avg_time, name, test_id)
end


function algo(arr::Vector{Vector{Int64}}, test_id::String, repetitions::Int)
	run_algo(arr, Seq.seq, "seq", test_id, repetitions)
	# run_algo(arr, Sgl_stc.sgl_stc, "sgl_stc", test_id, repetitions)
end # algo

end # module
